!!! FEATURE: Bind `ContentGraph` to workspace

kitsunet opened the pull request:
The ContentGraph is now always bound to a `WorkspaceName` and `ContentStreamId` combination.
The internal! ProjectionState of the ContentGraph is now the `ContentGraphFinder` which allows you to get an instance of the `ContentGraph` bound to a combination (usually by giving only the `WorkspaceName`).

The userland way to access the content graph is to ask the content repository:

`ContentRepository::getContentGraph(WorkspaceName)`


**Upgrade instructions**

The `WorkspaceName` must be passed along to get a subgraph instead of the internal `ContentStreamId`:

```diff
- $subgraph = $contentRepository->getContentGraph()->getSubgraph(
-     $workspace->currentContentStreamId,
-     $dimensionSpacePoint,
-     VisibilityConstraints::withoutRestrictions()
- );              
+ $subgraph = $contentRepository->getContentGraph($workspace->workspaceName)->getSubgraph(
+     $dimensionSpacePoint,
+     VisibilityConstraints::withoutRestrictions()
+ );
```

This pr presents an in between step. Its for example currently not easily possible to get the WorkspaceName of a `Node`, but that will be solved in a following pr see also https://github.com/neos/neos-development-collection/issues/4564.
In this intermediate step, one must ask the workspace finder first to resolve a workspace for content stream which then can be passed to  `getContentGraph` (see how `ContentRepositoryRegistry::subgraphForNode` handles it internally)

**Review instructions**

Followups: https://github.com/neos/neos-development-collection/issues/5038
Neos Ui adjustments: https://github.com/neos/neos-ui/pull/3775

kitsunet commented:
Ok legacy migration tests fail, I guess changes are needed there, I had done the obvious code adaptions, but probably that's not enough

mhsdesign commented on Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphFactory.php:
```diff
@@ -0,0 +1,64 @@
+
+    public function buildForWorkspace(WorkspaceName $workspaceName): ContentGraph
+    {
+        // FIXME: Should be part of this projection, this is forbidden
```
> note the big fat FIXME in both adapters factory implementations as those now require usage of a table that connects workspce to csId, and given that that is not there yet I hacked it, obviously thís needs to change

kitsunet replied:
I just realized this morning that I _could_ possibly move this resolving into the ContentGraphFinder which could have legitimate access to a workspace finder, just not sure I can easily resolve this dependency...
Either way we want this in the CG projection (adapters) anyways.

mhsdesign replied:
> Either way we want this in the CG projection (adapters) anyways.

yes... thesis is everything that is used for validating commands (soft constraints) must be now provided by the one cg projection so everything is more consistent and less flaky.

mhsdesign replied:
I think we can merge this first with these changes and followup by embedding the projection

-> https://github.com/neos/neos-development-collection/issues/5038

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/WorkspaceCommandHandler.php:
```diff
@@ -125,14 +126,19 @@ private function handleCreateWorkspace(
+            $command->workspaceName->value
+        ), 1505830958921);
+
         $baseWorkspace = $contentRepository->getWorkspaceFinder()->findOneByName($command->baseWorkspaceName);
```
maybe well get rid of this `WorkspaceFinder` here too someday ...

followup -> https://github.com/neos/neos-development-collection/issues/5038

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/NodeMove/NodeMove.php:
```diff
@@ -52,25 +53,22 @@ abstract protected function getInterDimensionalVariationGraph(): DimensionSpace\
     abstract protected function requireNodeAggregateToBeChild(
-        ContentStreamId $contentStreamId,
+        ContentGraphInterface $contentGraph,
         NodeAggregateId $childNodeAggregateId,
-        NodeAggregateId $parentNodAggregateId,
+        NodeAggregateId $parentNodeAggregateId,
         DimensionSpacePoint $dimensionSpacePoint,
-        ContentRepository $contentRepository,
     ): void;
```
i really love those signature changes ❤️ 
this makes the code inside much more easy to reason about ... like we only pass the `ContentGraph` and not the WHOLE cr :D

kitsunet committed:
TASK: ContentGraph bound to Workspace

The ContentGraph is now always bound to a
`WorkspaceName` and `ContentStreamId` combination.
The ProjectionState of the ContentGraph is now
the `ContentGraphFinder` which allows you to get
an instance of the `ContentGraph` bound to a
combination (usually by giving only the `WorkspaceName`).

The userland way is to use

`ContentRepository::getContentGraph(WorkspaceName)`

kitsunet commented on Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentSubgraph.php:
```diff
@@ -101,18 +87,21 @@
     public function __construct(
         private readonly ContentRepositoryId $contentRepositoryId,
+        /** @phpstan-ignore-next-line  */
+        private readonly WorkspaceName $workspaceName,
```
Not used anywhere right now but I figured we _need_ it for creating Nodes in the next changes that follow on this, so I thought it makes sense to add it already 🤷

bwaidelich commented on Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphFactory.php:
```diff
@@ -0,0 +1,64 @@
+ * @internal only used inside the
+ * @see ContentGraphFinder
+ */
+readonly final class ContentGraphFactory implements ContentGraphFactoryInterface
```
I wonder that the linters didn't seem to complain, but this is the order we use everywhere IIRC
```suggestion
final readonly class ContentGraphFactory implements ContentGraphFactoryInterface
```

bwaidelich commented on Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphFactory.php:
```diff
@@ -0,0 +1,64 @@
+        $tableName = strtolower(sprintf(
+            'cr_%s_p_%s',
+            $this->contentRepositoryId->value,
+            'Workspace'
```
Nitpick, but to save one needless character conversion:
```suggestion
            'workspace'
```
or even just use
```php
$tableName = sprintf('cr_%s_p_workspace', $this->contentRepositoryId->value);
```
like done in other places? The CR ID can only be lower case

kitsunet replied:
Yep, this was literally a last minute thing because the tests ran for me (case insensitive file system for the db table files -> need to move them) vs. case sensitive on GH actions. Didn't check if CR ID is lowercase but yeah lets fix it this way. it would become nicer when the workspace projection is part of the CG adapter anywys as then we would do the contentGraphTableNames->workspace() thing 😄

bwaidelich commented on Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphFactory.php:
```diff
@@ -0,0 +1,64 @@
+        $row = $this->client->getConnection()->executeQuery(
+            '
+                SELECT * FROM ' . $tableName . '
+                WHERE workspaceName = :workspaceName
```
Not sure if that's an improvement, so feel free to ignore
```suggestion
                WHERE workspaceName = :workspaceName
                LIMIT 1
```

bwaidelich commented on Neos.ContentGraph.DoctrineDbalAdapter/src/ContentGraphTableNames.php:
```diff
@@ -0,0 +1,44 @@
+ * Encapsulates table name generation for content graph tables
+ * @internal
+ */
+final class ContentGraphTableNames
```
Just wondering: In other implementations we just pass the `$tableNamePrefix` through and instead of:
```php
new Table($this->contentGraphTableNames->node(), ...
```
do something like
```php
new Table($this->tableNamePrefix . '_node', ...
```

Not a rhetoric question: do you think, it makes things easier to introduce a class for this? Because if so, we should maybe adjust other projections too

kitsunet replied:
Good question moving through a lot of this code I really was annoyed by adding these "magic" strings again and again with (well not for node but hierachyrelation for example) the risk of making a typo, I felt having this encapsulated would make it way easier to deal with refactorings. But that is certainly debatable. To me this looks way cleaner having not to rely on adding this formula of "$this->tableNamePrefix . '_node'" concat everywhere across multiple classes.

mhsdesign replied:
I also started adjusting what bastian noted and i also thought: Why always recreate the helper instance instead of passing it along:

https://github.com/kitsunet/neos-development-collection/pull/3

But it seems there are still a LOT of places using the legacy string syntax - for example the content graph projection will even use both ways currently (id 'deprecated' the second):

```
private readonly ContentGraphTableNames $contentGraphTableNames;
/** @deprecated */
private readonly string $tableNamePrefix;
```

Also the `NodeRecord` and `HierarchyRelation` active records mix the style (by using the new `DimensionSpacePointsRepository`)

Additionally i would definitely prefer a shortened name like `$tableNames` to simplify readably (we know we are scoped to the graph)

In the end im not against this refactoring - though it should have probably been separated ^^ - but it should be carried out consistently. Ill try to see if i can easily adjust the leftover cases tomorrow (see my pr).

mhsdesign replied:
Okay i refactored all the things in my pr. And found a typo `ContentGraphTableNames::hierachyRelation` vs `hierarchyRelation` :D

mhsdesign replied:
Last todo is to adjust the projection integrity violation thingy and make $tablePrefixName private

bwaidelich commented on Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php:
```diff
@@ -165,19 +158,22 @@ public function reset(): void
-        $contentGraph = $this->getState();
-        foreach ($contentGraph->getSubgraphs() as $subgraph) {
-            $subgraph->inMemoryCache->enable();
+        $contentGraphFinder = $this->getState();
+        /** @var ContentGraph $contentGraph */
+        foreach ($contentGraphFinder->getInstances() as $contentGraph) {
+            foreach ($contentGraph->getSubgraphs() as $subgraph) {
+                $subgraph->inMemoryCache->enable();
+            }
```
Not entirely new, but this code just made me a bit suspicious and I wonder if it is still needed when we switch to synchronous catchups

bwaidelich commented on Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentGraph.php:
```diff
@@ -126,76 +130,47 @@ public function findRootNodeAggregateByType(
         }
 
-        $rootNodeAggregate = $rootNodeAggregates->first();
-
-        if ($rootNodeAggregate === null) {
+        if (!$rootNodeAggregates->first()) {
             throw RootNodeAggregateDoesNotExist::butWasExpectedTo($nodeTypeName);
         }
 
-        return $rootNodeAggregate;
+        return $rootNodeAggregates->first();
```
Why did you change this? IMO the former code was slightly better because it does not re-fetch the first result which (in theory) could be changed in the meantime

kitsunet replied:
Good question, why did I change this^^ Lets change it back, I do not remember and looking at it makes no sense to me either.

bwaidelich reviewed (commented):
@kitsunet Thanks so much for putting a tremendous amount of time into this. I think it looks great already.
I just left some mostly nitpicky comments after a first quick pass mostly to dump my thoughts – no need to fix those (now).

One more important architectural choice is the stateful `ContentGraphFinder`.
If I get it right, it now needs a `getInstances()` method such that we can access (and reset the state of) all previously created CG instances in order to enable or disable their inMemory cache.
And then we need a `reset()` method to reset the instances for tests.

This is quite a lot of "burden" to an interface (and custom implementations) for some really low-level goals.
Maybe together we can come up with a slightly different mechanism to work with those runtime caches. But this is certainly no reason to block this one, we could always tweak it in a follow-up.

Great work!

kitsunet commented:
> @kitsunet Thanks so much for putting a tremendous amount of time into this. I think it looks great already. I just left some mostly nitpicky comments after a first quick pass mostly to dump my thoughts – no need to fix those (now).
> 
> One more important architectural choice is the stateful `ContentGraphFinder`. If I get it right, it now needs a `getInstances()` method such that we can access (and reset the state of) all previously created CG instances in order to enable or disable their inMemory cache. And then we need a `reset()` method to reset the instances for tests.
> 
> This is quite a lot of "burden" to an interface (and custom implementations) for some really low-level goals. Maybe together we can come up with a slightly different mechanism to work with those runtime caches. But this is certainly no reason to block this one, we could always tweak it in a follow-up.
> 
> Great work!

<3 Thank you!

Yes lets definitely have a look at those low level things, they were really an afterthought since I realized that definitely the tests do not like having the instances around. We could completely skip the cache, but then any cache within the CG or subgraph wouldn't be much use either because you get a fresh instance every time.

kitsunet committed:
Apply suggestions from code review

Co-authored-by: Bastian Waidelich <b.waidelich@wwwision.de>

mhsdesign commented on Neos.ContentRepository.Core/Classes/ContentGraphFinder.php:
```diff
@@ -0,0 +1,106 @@
+ * @see ContentGraphInterface
+ * @api only for read access during write operations and in services
+ */
+final class ContentGraphFinder implements ContentGraphFinderInterface
```
Maybe rather `ContentGraphProvider` ... i understand the relation to the `WorkspaceFinder` name wise but there are no `find*` methods.

kitsunet replied:
See comment below but totally not married to the name here, lets discuss what we find best and I change it.

mhsdesign commented on Neos.ContentRepository.Core/Classes/ContentGraphFinder.php:
```diff
@@ -0,0 +1,106 @@
+     * @throws WorkspaceDoesNotExist if there is no workspace with the provided name
+     * @throws ContentStreamDoesNotExistYet if the provided workspace does not resolve to an existing content stream
+     */
+    public function fromWorkspaceName(WorkspaceName $workspaceName): ContentGraphInterface
```
i think `for*` or `provideForWorkspaceName` (like the naming from the `WorkspaceProvider` makes more sense?

kitsunet replied:
riiiight yeah I actually wanted to do that, if we treat this as a finder it should even be `get` ? unless we would add a check if the workspace exist and return null? Might be an idea and then do `findBy`

mhsdesign commented on Neos.ContentRepository.Core/Classes/Projection/ContentGraph/ContentGraphInterface.php:
```diff
@@ -158,4 +148,8 @@ public function getDimensionSpacePointsOccupiedByChildNodeName(
      */
     public function countNodes(): int;
+
+    public function getWorkspaceName(): WorkspaceName;
+
+    public function getContentStreamId(): ContentStreamId;
```
```suggestion
    /** The workspace this content graph is operating on */
    public function getWorkspaceName(): WorkspaceName;

    /** @internal The content stream id where the workspace name points to for this instance */
    public function getContentStreamId(): ContentStreamId;
```

and maybe add a

> see live cycle of content graph in ContentGraphProvider

mhsdesign commented on Neos.ContentRepository.Core/Classes/ContentGraphFinder.php:
```diff
@@ -0,0 +1,106 @@
+ * @see ContentGraphInterface
+ * @api only for read access during write operations and in services
+ */
+final class ContentGraphFinder implements ContentGraphFinderInterface
```
This thing here should get a nice explanation about the live cycle and cache behaviour of the ContentGraph to answer all questions ...

maybe to better explain caching and state the `ContentGraphFactoryInterface` should rather have a `getCurrentContentStreamIdForWorkspace` so the workspace name is manually resolved here and then put into `buildForWorkspaceAndContentStream`? That _could_ make things more obvious and that way we put all the heavy lifting here.

kitsunet replied:
Yes that could definitely be an option

bwaidelich replied:
Some doc comments would surely help.
Nevertheless, I feel that we should lift the instance coordination to a higher level.
E.g. in Neos-land the new `Workspace` model will probably become the defacto standard way to access a content (sub)graph. So maybe it could keep the instances.

```php
foreach ($contentGraphFinder->getInstances() as $contentGraph) {
    foreach ($contentGraph->getSubgraphs() as $subgraph) {
        $subgraph->inMemoryCache->enable();
    }
}
```
This code is quite fragile (it's not new!) because it leaks a whole lot of internals (e.g. it relies on the `$subgraph` to have a field `inMemoryCache` etc).

IMO even a static `ContentSubgraphWithRuntimeCaches::resetRuntimeCaches()` or `ContentSubgraphWithRuntimeCaches::resetRuntimeCachesForInstance(ContentSubgraphIdentity $identity)` might work

kitsunet replied:
Yeah when I found this code I didn't like it at all, but didn't spontaneously have a satisfactory idea. Not sure I like the static either.

mhsdesign replied:
Kinda related to https://github.com/neos/neos-development-collection/pull/4955

if the behaviour dindt change, which i think it didnt, and we dont block any paths to a better future with this pr i would suggest to move this discussion to a separate thread to not block this pr.

Dont get me wrong, the runtime cache thing must be addressed but seeing this is a fundamental change for many prs to come i rather fix it later ;)

bwaidelich replied:
> if the behaviour dindt change, which i think it didnt, and we dont block any paths to a better future with this pr
> i would suggest to move this discussion to a separate thread to not block this pr.

This adds public methods to userland code, that are not meant to be used by users. That's why it got slightly more important to get it right.

> this is a fundamental change for many prs to come i rather fix it later ;)

Absolutely! And as mentioned elsewhere: I don't mean to block this one, we can certainly iterate via follow-ups!

kitsunet replied:
We _could_ declare the CGFinder internal and relegate all user land cdoe to please only ever use CR->getContentGraph() to safeguard a bit for now, while we iterate 😆

bwaidelich replied:
yep yep. don't worry about it too much for now!

mhsdesign replied:
Ah a see sorry yes `ContentGraphFinder::reset` and `ContentGraphFinder::getInstances` are indeed new.
Previously we only exposed via the doctrine adapter `\Neos\ContentGraph\DoctrineDbalAdapter\...\ContentGraph::getSubgraphs` but we didnt have `public` methods declared inside of the core that are not meant to be used.
Declaring this thing internal would be my best bet too for now ;)

kitsunet committed:
Add comments to new ContentGraph methods

Co-authored-by: Marc Henry Schultz <85400359+mhsdesign@users.noreply.github.com>

bwaidelich commented on Neos.ContentRepository.Core/Classes/ContentGraphFinderInterface.php:
```diff
@@ -0,0 +1,36 @@
+ * @see ContentGraphInterface
+ * @internal
+ */
+interface ContentGraphFinderInterface extends ProjectionStateInterface
```
Why do we need that interface?
You added `reset()` only to the single `ContentGraphFinder` implementation and call it from the code (e.g. in `WorkspaceCommandHandler::resetContentGraphCache()` and in Behat traits.
We should instead reference the interface (and add required methods) or remove it (which I would prefer)

kitsunet replied:
I think I originally thought to separate userland from internal implementation parts by annotating inteface or implementation respectively, BUT that is moot as the userland part is ContentRepository::getContentGraph(WorkspaceName) and that's it, I will drop it.

kitsunet commented on Neos.ContentRepository.Core/Classes/ContentGraphFinder.php:
```diff
@@ -0,0 +1,106 @@
+     * @return ContentGraphInterface
+     * @internal
+     */
+    public function fromWorkspaceNameAndContentStreamId(WorkspaceName $workspaceName, ContentStreamId $contentStreamId): ContentGraphInterface
```
So this is ONLY used in tests, and TBH again last minute addition because I didn't see an easy way out. We need to talk about this. The behat tests do some ugly stuff internally...

mhsdesign replied:
Yes i know the test use nice reflection and stuff ^^ i started a conversation here about this https://github.com/neos/neos-development-collection/issues/4634

kitsunet commented on Neos.ContentRepository.Core/Classes/Projection/ContentGraph/ContentGraphInterface.php:
```diff
@@ -158,4 +148,8 @@ public function getDimensionSpacePointsOccupiedByChildNodeName(
      * @internal only for consumption in testcases
      */
     public function countNodes(): int;
```
This one is a very odd duck now as it is only used for tests AND operates globally (so doesn't take the workspace this is bound to into account). I am open for suggestions. We could make it operate on the workspace / content stream and see if that works for the tests too?!

mhsdesign replied:
Yeah or at the very least declare it as experimental and tell that it operates globally ... would be definitely fine to address this in a followup ;)

kitsunet replied:
Mmm, adapting it to operate on bound content stream is not so trivial, we then need to join on hierachy relation and I fear this muddles with the intend of this being super low level and catching possible mismatches in the projection as well 🤔

mhsdesign replied:
Yes dont worry i think we can open a followup issue for this and discuss wether we like this and how we like this ;)
(For example it _could_ be a method directly on the `ContentGraphFinder`)

kitsunet committed:
Code cleanup and documentation

kitsunet committed:
Fix linter issue

mhsdesign commented on Neos.ContentRepository.LegacyNodeMigration/Tests/Behavior/Bootstrap/FeatureContext.php:
```diff
@@ -123,10 +124,14 @@ public function iRunTheEventMigration(string $contentStream = null): void
-        $contentGraph = $this->currentContentRepository->getContentGraph();
-        $nodeFactory = (new \ReflectionClass($contentGraph))
+        $contentGraphFinder = $this->currentContentRepository->projectionState(\Neos\ContentRepository\Core\ContentGraphFinder::class);
+        // FIXME: Dirty
+        $contentGraphFactory = (new \ReflectionClass($contentGraphFinder))
+            ->getProperty('contentGraphFactory')
+            ->getValue($contentGraphFinder);
+        $nodeFactory = (new \ReflectionClass($contentGraphFactory))
             ->getProperty('nodeFactory')
-            ->getValue($contentGraph);
+            ->getValue($contentGraphFactory);
```
Nice reflection and stuff ^^ fyi i started a conversation here about this https://github.com/neos/neos-development-collection/issues/4634

kitsunet replied:
This is so bad, but it's just an extension of what was there before, we need to solve this nicely, and if it's just for tests then we should change how we test.

bwaidelich replied:
We should probably just create some CR Service for the behat tests – that has access to all the `ContentRepositoryServiceFactoryDependencies` including the `propertyConverter` that is the only thing we're extracting here.

But we can certainly solve that in a follow-up!

bwaidelich replied:
Actually this is much more an issue of the `LegacyMigrationService` in this case.
We can definitely fix that and make it easier to be used from tests!

mhsdesign commented on Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/CRTestSuiteRuntimeVariables.php:
```diff
@@ -160,8 +161,13 @@ public function visibilityConstraintsAreSetTo(string $restrictionType): void
     {
-        return $this->currentContentRepository->getContentGraph()->getSubgraph(
-            $this->currentContentStreamId,
+        $contentGraphFinder = $this->currentContentRepository->projectionState(ContentGraphFinder::class);
+        $contentGraphFinder->reset();
+        if (isset($this->currentContentStreamId)) {
+            return $contentGraphFinder->getByWorkspaceNameAndContentStreamId($this->currentWorkspaceName, $this->currentContentStreamId)->getSubgraph($this->currentDimensionSpacePoint, $this->currentVisibilityConstraints);
+        }
+
+        return $contentGraphFinder->getByWorkspaceName($this->currentWorkspaceName)->getSubgraph(
             $this->currentDimensionSpacePoint,
             $this->currentVisibilityConstraints
```
that seems kindof crazy and too intelligent ^^ .. is the `currentContentStreamId` set in the majority of the test or if not all? Then the actual code below is just dummy.

kitsunet replied:
There are tests that only refer to be in the workspace "XYZ" and it's current content stream, in that case only the workspace could be set.

mhsdesign replied:
I see there are 5 different step syntaxes to change the workspace and or content stream id.

| Step | Modifies | Usages |
|--------|--------|--------|
| `I am in workspace "live"` | `currentWorkspaceName` | 8 |
| `I am in content stream "cs-id"` or <br> `I am in content stream "cs-id" and dimension space point {}`  | `currentContentStreamId` | 113 |
| `I am in the active content stream of workspace "live"` or <br> `I am in the active content stream of workspace "live" and dimension space point {}`  | `currentWorkspaceName` & `currentContentStreamId` | 788 | 

We hardly use the logic you introduced - only 8 times - as otherwise the `currentContentStreamId` will be set.

Also problematic, as you pointed out in [slack](https://neos-project.slack.com/archives/C04PYL8H3/p1714846896689819), is that the step `I am in content stream "cs-id" ...` will no longer be supported really as api ...

I dont know how what the plan is cc @nezaniel

nezaniel replied:
If we don’t communicate content stream ids any more, we don’t need them in the test runtime variables either

kitsunet replied:
Right, but that is a change for another day I guess

nezaniel replied:
Absolutely, yes.
Recommended way forward:
1) merge this
2) remove the remainig usages of test runtime variable currentContentSteamId in PHP, if any
3) rewrite all the test "given"s

mhsdesign replied:
moved https://github.com/neos/neos-development-collection/issues/5034

bwaidelich committed:
Tweak ContentGraphFinder runtime cache

bwaidelich committed:
Merge branch 'feature/content-graph-bound' of https://github.com/kitsunet/neos-development-collection into feature/content-graph-bound

# Conflicts:
#	Neos.ContentRepository.Core/Classes/ContentGraphFinder.php

bwaidelich commented:
@kitsunet I took the freedom to push our changes from yesterday (24cfa06d01200ca7f19d7243b4fab59cb6ebca0a) to this PR. I don't know if tests are still green because there is a conflict that I wasn't able to resolve

mhsdesign commented on Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php:
```diff
@@ -234,26 +223,14 @@ public function getCheckpointStorage(): DbalCheckpointStorage
     {
-        $contentGraph = $this->getState();
-        foreach ($contentGraph->getSubgraphs() as $subgraph) {
-            $subgraph->inMemoryCache->disable();
-        }
+        $this->getState()->forgetInstances();
     }
```
this changes behaviour as in previously an instance of a `$subgraph` held by a cr user would get its cache flushed. Now not anymore.
This is not certainly a dealbreaker but we should most definitely define the behaviour as we expect it either by test and or by documentation so we can just copy and pasta it to slack ;)

bwaidelich replied:
good point, this _is_ a dealbreaker IMO – probably it makes sense to call this something else (e.g. `resetRuntimeCaches`) and then reset the inMemoryCache (we should not _disable_ it).

The main goal of this change was not to expose all those nasty internals such that we can refactor more easily.
In fact `inMemoryCache` is now only public because it is still used by the nasty code of the [frontend controller](https://github.com/neos/neos-development-collection/blob/9.0/Neos.Neos/Classes/Controller/Frontend/NodeController.php#L314-L333) (which I plan to clean up)

But we should make sure that:
```php
$subGraph->find*(...);
$contentRepository->handle(...);
$subGraph->find*(...);
```
behaves as expected, i.e. the second subgraph interaction contains the changes.

Lets see if the tests fail – and if they don't we should add scenarios that reflect the above!

mhsdesign replied:
👍 except that the code you showed is doomed in some scenarios either way

```php
$subGraph->find*(...);
$contentRepository->handle(DiscardChanges(...));
$subGraph->find*(...);
```

will not work on a more fundamental way, because the content stream id changes and the one bound to the instance of a $subgraph or $contentgraph contains the old one.
Im not sure what the behaviour will be at this point but likely it will just return nothing as the old content stream is dead.

bwaidelich replied:
We'll have to think about that one, but once we no longer refer content streams it _should_ be possible I think

mhsdesign replied:
i thought the assumption was that the contentStreamId will be resolved ON construction of a subgraph (or even before that, on construction of a contentGraph and then `getSubgraph` will just forward it).
This is imo an important aspect that we need to figure out.

Basically the question of joining at runtime ALWAYS or caching the csId by setting it on construction.
Especially since there will be `contentGraph::getContentStreamId` this question is important.

But i thought we agreed upon that the subgraph / or content graph will NOT mutate and operate on the state it was created even if this means that it will fail because the cs id is no longer existing.

so this will be the safest, and if we embrace this pattern we wont need to clear the in memory cache ^^

```php
$contentRepository->getContentGraph()->getSubgraph()->find*(...);
$contentRepository->handle(DiscardChanges(...));
$contentRepository->getContentGraph()->getSubgraph()->find*(...);
```

kitsunet replied:
Seeing the one broken test says something has chnaged and I think it's somehow related to these caches, although I can't really figure out what yet.

kitsunet replied:
> caching the csId by setting it on construction

IMHO we need that. A ContentGraph and by extension it's subgraphs are bound to a workspace/cs combo. I don't see way around that.

bwaidelich replied:
yes, I think you're right and I mixed up things.
What I meant to say is:
```php
$contentRepository->getContentGraph()->getSubgraph()->find*(...);
$contentRepository->handle(MoveNodeAggregate(...));
$contentRepository->getContentGraph()->getSubgraph()->find*(...);
```
The second call should not just return the runtime-cached results.

For the (more unlikely) case that you discard/publish/rebase during one process it would be fine to require re-fetching of the subgraph IMO – but the old instance should not just return previously cached results in that case either

kitsunet replied:
Indeed!

mhsdesign replied:
i discussed this matter again with Bernhard and we agree that having a subgraph instance in your hand after doing major changes like Discard or Publish it must be refetched as otherwise the behaviour is undefined (we werent even sure if the old cs-id still exists then in the graph projection)

BUT for simple changes like setting a property a cr user will expect the in memory cache to be flushed (not disabled!).

mhsdesign replied:
ups i was trying to add it here but then in struck me yes no sure that wont work because we dont know about this implementation detail and would have to hack:

```php
public function forgetInstances(): void
{
    // we also clear the in memory caches for the case someone still holds an instance to a $subgraph
    // and the subgraph should continue to work when fetching an updated node after changing its properties.
    // if the workspace to content stream relation changes using the subgraph or content graph is undefined behaviour!
    foreach ($this->contentGraphInstances as $contentGraph) {
        if (method_exists($contentGraph, 'getSubgraphs')) {
            foreach ($contentGraph->getSubgraphs() as $subgraph) {
                $subgraph->inMemoryCache->disable();
            }
        }
    }

    $this->contentGraphInstances = [];
}
```

kitsunet replied:
This discussion should go to another ticket? Because it's now completely unrelated to this change as the inMemoryCache will be a new instance via markStale() so for write side this is mitigated, and everythign else is another story

mhsdesign replied:
As discussed we are ok with this regression and fix this in a followup: https://github.com/neos/neos-development-collection/issues/5039

mhsdesign commented on Neos.ContentRepository.TestSuite/Classes/Behavior/Features/Bootstrap/ProjectedNodeTrait.php:
```diff
@@ -265,7 +266,15 @@ public function iExpectThisNodeToExactlyInheritTheTags(string $tagList): void
+        } else {
+            $contentGraph = $this->currentContentRepository->getContentGraph($this->currentWorkspaceName);
+        }
+        // FIXME: query->workspaceName this _might_ be wrong as the query could use a different workspace/contentstream
+        $this->currentNode = $query($contentGraph);
```
i dont fully understand the todo

kitsunet replied:
removing, doesn't matter anymore, I think it was more a marker for me during refactoring.

kitsunet committed:
Merge remote-tracking branch 'origin/9.0' into feature/content-graph-bound

kitsunet commented on Neos.ContentGraph.DoctrineDbalAdapter/src/DoctrineDbalContentGraphProjection.php:
```diff
@@ -237,13 +230,7 @@ public function getState(): ContentGraphFinder
-                $subgraph->inMemoryCache->disable();
-            }
-        }
+        $this->getState()->forgetInstances();
```
Ok this specifically now breaks the one failing test, as before we only purged the inMemory cache but not the finder instances. It's amazing that of all the tests we just have one failing because this here is a fundamental break. 

The `\Neos\ContentRepository\Core\EventStore\EventPersister::publishEvents()` does call markStale() but that means the contentStreamIdOverrides are gone as well so during a workspace publish/rebase operation some events (everfy after the first) might apply to the wrong content stream as the overriden "special" instance was purged and will then automatically be replaced by a automatically resolved (by workspace) one which has the wrong content stream.

We _could_ separate the override instances but that seems messy as well?

mhsdesign replied:
Wow very good catch. This is indeed a flaw that should be guarded against.

> We _could_ separate the override instances but that seems messy as well?

i would actually prefer some `$this->isInOverrideState` flag than how you implemented it sneakily ^^

kitsunet committed:
Separate CommandHandlingDependencies

kitsunet committed:
Fix phpcs errors

kitsunet committed:
Fix lint errors

bwaidelich commented on Neos.ContentRepository.Core/Classes/CommandHandlingDependencies.php:
```diff
@@ -0,0 +1,76 @@
+use Neos\ContentRepository\Core\SharedModel\Workspace\WorkspaceName;
+
+/**
+ * @internal only command handlers
```
only command handlers _what_? :)

bwaidelich commented on Neos.ContentRepository.Core/Classes/CommandHandlingDependencies.php:
```diff
@@ -0,0 +1,76 @@
+<?php
+
```
```suggestion

declare(strict_types=1);

```

bwaidelich commented on Neos.ContentRepository.Core/Classes/CommandHandlingDependencies.php:
```diff
@@ -0,0 +1,76 @@
+/**
+ * @internal only command handlers
+ */
+class CommandHandlingDependencies
```
```suggestion
final class CommandHandlingDependencies
```

bwaidelich commented on Neos.ContentRepository.Core/Classes/Factory/ContentRepositoryFactory.php:
```diff
@@ -100,7 +101,7 @@ public function getOrBuild(): ContentRepository
                 $this->projectionFactoryDependencies->contentDimensionSource,
                 $this->userIdProvider,
-                $this->clock,
+                $this->clock
```
Please leave those trailing commas in so we don't create multi-line-diffs when adding new arguments
```suggestion
                $this->clock,
```

bwaidelich commented on Neos.ContentRepository.Core/Classes/Factory/ContentRepositoryFactory.php:
```diff
@@ -120,11 +121,12 @@ public function getOrBuild(): ContentRepository
             $this->getOrBuild(),
             $this->buildEventPersister(),
-            $this->projectionsAndCatchUpHooks->projections,
+            $this->projectionsAndCatchUpHooks->projections
```
```suggestion
            $this->projectionsAndCatchUpHooks->projections,
```

bwaidelich commented on Neos.ContentRepository.Core/Classes/Factory/ContentRepositoryServiceFactoryDependencies.php:
```diff
@@ -46,7 +46,7 @@ private function __construct(
         // we don't need CommandBus, because this is included in ContentRepository->handle()
         public EventPersister $eventPersister,
-        public Projections $projections,
+        public Projections $projections
```
```suggestion
        public Projections $projections,
```

bwaidelich commented on Neos.ContentRepository.Core/Classes/Factory/ContentRepositoryServiceFactoryDependencies.php:
```diff
@@ -57,7 +57,7 @@ public static function create(
         ContentRepository $contentRepository,
         EventPersister $eventPersister,
-        Projections $projections,
+        Projections $projections
```
```suggestion
        Projections $projections,
```

bwaidelich commented on Neos.ContentRepository.Core/Classes/Factory/ContentRepositoryServiceFactoryDependencies.php:
```diff
@@ -70,7 +70,7 @@ public static function create(
             $contentRepository,
             $eventPersister,
-            $projections,
+            $projections
```
```suggestion
            $projections,
```

bwaidelich commented on Neos.ContentRepository.Core/Classes/Feature/Common/NodeVariationInternals.php:
```diff
@@ -291,33 +269,20 @@ protected function collectNodePeerVariantsThatWillHaveBeenCreated(
             // check the siblings succeeding in the origin dimension space point
             foreach ($originSiblings as $originSibling) {
-                $variantSibling = $variantSubgraph->findNodeById($originSibling->nodeAggregateId);
+                $variantSibling = $contentGraph->getSubgraph($variantDimensionSpacePoint, VisibilityConstraints::withoutRestrictions())->findNodeById($originSibling->nodeAggregateId);
```
Just wondering: Why did you inline this?
I found the previous version easier to read and it didn't refetch the same instance in every loop

kitsunet replied:
Yeah no good reason, this is a leftover from the original Adapter change I think, were inlining made more sense.

bwaidelich commented on Neos.ContentRepository.Core/Classes/Feature/ContentStreamCommandHandler.php:
```diff
@@ -130,14 +129,13 @@ private function handleReopenContentStream(
 
-        $sourceContentStreamVersion = $contentRepository->getContentStreamFinder()
-            ->findVersionForContentStream($command->sourceContentStreamId);
+        // TOOD: THis is not great
```
```suggestion
        // TODO: This is not great
```

and: why is it not great?

bwaidelich commented on Neos.ContentRepository.Core/Classes/Feature/ContentStreamCommandHandler.php:
```diff
@@ -105,11 +104,11 @@ private function handleCloseContentStream(
     private function handleReopenContentStream(
         ReopenContentStream $command,
-        ContentRepository $contentRepository
+        CommandHandlingDependencies $commandHandlingDependencies
```
```suggestion
        CommandHandlingDependencies $commandHandlingDependencies,
```

bwaidelich commented on Neos.ContentRepository.Core/Classes/Feature/ContentStreamCommandHandler.php:
```diff
@@ -85,11 +84,11 @@ private function handleCreateContentStream(
     private function handleCloseContentStream(
         CloseContentStream $command,
-        ContentRepository $contentRepository
+        CommandHandlingDependencies $commandHandlingDependencies
```
```suggestion
        CommandHandlingDependencies $commandHandlingDependencies,
```

bwaidelich commented on Neos.ContentRepository.Core/Classes/Feature/NodeModification/NodeModification.php:
```diff
@@ -152,7 +150,7 @@ private function splitPropertiesToUnsetByScope(PropertyNames $propertiesToUnset,
 
         return array_map(
-            fn(array $propertyValues): PropertyNames => PropertyNames::fromArray($propertyValues),
+            static fn(array $propertyValues): PropertyNames => PropertyNames::fromArray($propertyValues),
```
not related to your change, but since you adjusted this line I would suggest to change this to:
```suggestion
            PropertyNames::fromArray(...),
```

bwaidelich commented on Neos.ContentRepository.Core/Classes/Projection/ContentGraph/ContentGraphInterface.php:
```diff
@@ -40,31 +41,28 @@ interface ContentGraphInterface extends ProjectionStateInterface
     ): ContentSubgraphInterface;
 
     /**
+     * @throws RootNodeAggregateDoesNotExist
```
again, not totally new. But the ordering of the annotations is weird

bwaidelich commented on Neos.ContentRepository.Core/Classes/Factory/ContentRepositoryFactory.php:
```diff
@@ -62,6 +62,7 @@ public function __construct(
             $contentDimensionSource,
             $contentDimensionZookeeper
         );
+        $propertyConverter = new PropertyConverter($propertySerializer);
```
You added this variable but don't use it
```suggestion
```

bwaidelich reviewed (commented):
I love it, so many issues are solved now.

I added some more comments (some nitpicky ones you can just skip over for now)

kitsunet committed:
Apply suggestions from code review

Co-authored-by: Bastian Waidelich <b.waidelich@wwwision.de>

mhsdesign commented:
FYI we have a MAJOR conflict with #5018 see https://github.com/neos/neos-development-collection/pull/5018#issuecomment-2102199661

kitsunet commented:
Oooof, again :(

I mean before operating the query changes out I rather fix the conflicts I guess, IDK

kitsunet committed:
Some cosmetic fixes

mhsdesign commented:
> Oooof, again :(

half as bad i just checked with Bernhard and he said this pr can go first and he will solve the conflicts, which are not thaaat bad ;)

mhsdesign commented on Neos.ContentRepository.Core/Classes/CommandHandlingDependencies.php:
```diff
@@ -0,0 +1,81 @@
+    {
+    }
+
+    public function handle(CommandInterface $command): CommandResult
```
maybe rather `handleNestedCommand`

mhsdesign replied:
or `handleSubCommand` or `delegate` but it doesnt really matter as its only used by the workspace command handler and we already discussed that it should rather be a service or such but is too overpowered :D so we can keep `handle`:D

mhsdesign commented:
Documented all followups here ;) https://github.com/neos/neos-development-collection/issues/5038

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/NodeDuplication/NodeDuplicationCommandHandler.php:
```diff
@@ -52,6 +54,11 @@ public function __construct(
     ) {
     }
 
+    protected function getContentGraph(WorkspaceName $workspaceName, CommandHandlingDependencies $commandHandlingDependencies): ContentGraphInterface
+    {
+        return $commandHandlingDependencies->getContentGraph($workspaceName);
+    }
```
why were these kind of methods added to multiple command handler? i think its a leftover piece?

kitsunet replied:
removed

mhsdesign committed:
TASK: Declare `ContentGraphFinder` as internal and pass whole CR around instead

mhsdesign committed:
TASK: Try to pass one `ContentGraphTableNames` instance along instead of magic strings

in some places we still use the old and new way.

mhsdesign committed:
TASK: Refactor `HierarchyRelation` active record to use `ContentGraphTableNames`

mhsdesign committed:
TASK: Refactor `NodeRecord` active record to use `ContentGraphTableNames`

mhsdesign committed:
TASK: Remove string based `tableNamePrefix` from `DoctrineDbalContentGraphProjection`

mhsdesign committed:
TASK: Rename `$contentGraphTableNames` to `$tableNames`

mhsdesign committed:
TASK: Remove dead code from ProjectionContentGraph

mhsdesign committed:
TASK: Use `ContentGraphTableNames` in `ProjectionContentGraph`

mhsdesign committed:
TASK: Fix typo in `ContentGraphTableNames::hierarchyRelation`

kitsunet commented on Neos.ContentRepository.Core/Classes/ContentGraphFinder.php:
```diff
@@ -22,12 +22,11 @@
+ * The API way of accessing a ContentGraph is via ContentRepository::getContentGraph()
  *
- * @api Not for userland code, only for read access during write operations and in services
+ * @internal User land code should not use this directly.
```
Now that the whole override thing is gone from here I would have actually left it API as it's a projection state like the others and it seems weird to have one internal and the rest API in there.

mhsdesign replied:
Well this is just intermediate step, i thought we want to make the content graph projection first level citizen and thus well soon drop this thing again.

kitsunet replied:
That would b e something to discuss today

kitsunet replied:
Maybe later make api, after we saw how it goes

kitsunet committed:
Remove unused and unuseful getContentGraph methods

kitsunet committed:
Merge pull request #3 from mhsdesign/task/refactorFurtherToContentGraphTableNames

TASK: Try to pass one `ContentGraphTableNames` instance along instead…

kitsunet committed:
Use TableNames in ProjectionIntegrityViolationDetector

kitsunet commented:
Hah, net minus lines, nice!

mhsdesign committed:
TASK: Throw `WorkspaceDoesNotExist` exception when calling `getContentGraph`

mhsdesign commented:
UI Adjustments: https://github.com/neos/neos-ui/pull/3775

mhsdesign committed:
TASK: Adjust executed exception to `WorkspaceDoesNotExist` in tests

mhsdesign committed:
TASK: Remove obsolete `getSubgraphs` method

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php:
```diff
@@ -696,7 +671,7 @@ protected function requireNodeAggregateToOccupyDimensionSpacePoint(
         if (!$nodeAggregate->occupiesDimensionSpacePoint($originDimensionSpacePoint)) {
             throw new DimensionSpacePointIsNotYetOccupied(
-                'Dimension space point ' . json_encode($originDimensionSpacePoint)
+                'Dimension space point ' . json_encode($originDimensionSpacePoint, JSON_THROW_ON_ERROR)
```
i guess `JSON_PARTIAL_OUTPUT_ON_ERROR` would be better but this is only for linting purposes so yolo

mhsdesign replied:
```suggestion
                'Dimension space point ' . json_encode($originDimensionSpacePoint, JSON_PARTIAL_OUTPUT_ON_ERROR)
```

i guess `JSON_PARTIAL_OUTPUT_ON_ERROR` would be better but this is only for linting purposes so yolo

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/ContentStreamCommandHandler.php:
```diff
@@ -130,14 +129,13 @@ private function handleReopenContentStream(
 
-        $sourceContentStreamVersion = $contentRepository->getContentStreamFinder()
-            ->findVersionForContentStream($command->sourceContentStreamId);
+        // TODO: This is not great
```
```suggestion
```

Huuuch a todo i did no know about?
If its about the `getContentStreamFinder` thing that we should better not use i have a followup planned ^^ but else what is this about?

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/ContentStreamCommandHandler.php:
```diff
@@ -130,14 +129,13 @@ private function handleReopenContentStream(
 
-        $sourceContentStreamVersion = $contentRepository->getContentStreamFinder()
-            ->findVersionForContentStream($command->sourceContentStreamId);
+        // TODO: This is not great
```
```suggestion
```

Huuuch a todo i did no know about?
If its about the `getContentStreamFinder` thing that we should better not use i have a followup planned ^^ but else what is this about?

kitsunet replied:
Lost in the history of this change, at least in it's current state it's fine. Maybe I had a markStale or somthing in there at some point.

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/Common/ConstraintChecks.php:
```diff
@@ -712,7 +687,7 @@ protected function requireNodeAggregateToNotOccupyDimensionSpacePoint(
         if ($nodeAggregate->occupiesDimensionSpacePoint($originDimensionSpacePoint)) {
             throw new DimensionSpacePointIsAlreadyOccupied(
-                'Dimension space point ' . json_encode($originDimensionSpacePoint)
+                'Dimension space point ' . json_encode($originDimensionSpacePoint, JSON_THROW_ON_ERROR)
```
```suggestion
                'Dimension space point ' . json_encode($originDimensionSpacePoint, JSON_PARTIAL_OUTPUT_ON_ERROR)
```

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/ContentStreamCommandHandler.php:
```diff
@@ -65,10 +66,8 @@ public function handle(CommandInterface $command, ContentRepository $contentRepo
-        ContentRepository $contentRepository
+        CreateContentStream $command
     ): EventsToPublish {
-        $this->requireContentStreamToNotExistYet($command->contentStreamId, $contentRepository);
```
why have all the two `requireContentStreamToNotExistYet` checks and the method itself been removed? (see also in `handleCloseContentStream`)

mhsdesign replied:
Ah i see because you also removed `ContentStreamFinder::hasContentStream` ... but this does not make the image clearer to me?

kitsunet replied:
Well no, I removed it because it was no longer in use after removing the calls, and I removed the calls, because at this point we want to rather look at workspaces, BUT I guess this case here is valid 🤷 I'll add it back

mhsdesign commented on Neos.Neos/Classes/Fusion/Cache/ContentCacheFlusher.php:
```diff
@@ -64,6 +65,8 @@ public function __construct(
     /**
      * Main entry point to *directly* flush the caches of a given NodeAggregate
      *
+     * FIXME workspaceName instead of contentStreamId
+     *
```
```suggestion
```

this is too far up, i would miss deleting this 😅

kitsunet replied:
It's about the param? Where else should it go without destroying the param block

mhsdesign replied:
idk ill take care of this, dont worry about having to annotate it all ;)

mhsdesign commented on Neos.Neos/Classes/AssetUsage/Dto/AssetIdAndOriginalAssetId.php:
```diff
@@ -9,6 +9,7 @@
 /**
  * @internal
  */
+
```
```suggestion
```

imo correct white space

kitsunet replied:
But this file wasn't even changed in this PR?

kitsunet replied:
Wait it was, why was this whitespace added in the first place, not by me IMHO.

mhsdesign reviewed (approved):
Wow. What a blast. As bastian had one last go through i also found some mininini things (just apply and merge :D)

Additionally i managed to get the Neos Ui e2e tests running in addition to the Neos ui adjustments i prepared.

So i consider this absolutely ready and it can be merged this evening!

kitsunet committed:
Apply suggestions from code review

Co-authored-by: Marc Henry Schultz <85400359+mhsdesign@users.noreply.github.com>

kitsunet committed:
Remove unnecessary empty line

Co-authored-by: Marc Henry Schultz <85400359+mhsdesign@users.noreply.github.com>

kitsunet committed:
Re-add content stream does not exist check

kitsunet commented:
Note for later:
Interesting that removal of the "doesNotExistYet" check did not lead any test to fail. Might be worth adding a test that checks that handling a command that tries to create an already existing ContentStream actually fails with the (re-added) exception.

mhsdesign commented:
Wow now its finally done. Thx for the plenty adjustments and your motivation.

@bwaidelich do you want to +1 as well?

bwaidelich committed:
Don't expose ContentGraphTableNames tableNamePrefix

bwaidelich commented:
@mhsdesign being trigger-happy again without even waiting for the checks to pass :)

mhsdesign commented:
Haha lol idk this was automerge? 😂
